### DISCUSSION TITLE
Time Series Online Learning

### DISCUSSION THREAD

**Discussion by ngupta23** [pycaret/pycaret#1767] [FAQs]
# Background

In many time series problems, we have to continuously train a model at regular intervals using new data that is available from the time the model was last trained. For example, we may have streaming data where we may want to continually model the stream using a time series model. In other cases, we may be on a monthly model training schedule (e.g. sales forecasting) and may want to update our model with the new "actual" data point that becomes available every month. In these types of problems, it is often faster to use the original model as a starting point and only update the model parameters based on the newly available data points (as opposed to retraining the model on the entire data set). This type of problem is called `Online Learning`. 

# Solution

Luckily, this kind of problem can now be solved with `pycaret` in conjunction with `sktime`. Let's see how. We will take the `airline` data set as an example. This data set contains data from Jan-1949 till Dec-1960. We will reserve all the data points except the last one (Dec-1960) to train the initial model. We will use the last data point to show how online learning can be performed once the initial model has been developed.

```python
#### Install library (one time) ----
!pip install pycaret-ts-alpha -U

#### Load data ----
from pycaret.datasets import get_data
y = get_data("airline")

#### Reserve data for online learning (last data point) ----
y_for_later = y[-1:]

#### Original model trained on all but the last data point ----
y = y[:-1] 

#### Setup experiment ----
from pycaret.internal.pycaret_experiment import TimeSeriesExperiment
exp = TimeSeriesExperiment()
exp.setup(data=y, fh=12, session_id=42)

#### Initial Training ----
model = exp.create_model("ets")
exp.plot_model(model)
```

![image](https://user-images.githubusercontent.com/33585645/138790140-be94f887-e878-4c15-9e06-3b3bf2131bc8.png)

If we are happy with this, we can finalize the model so we can make future predictions.

```python
#### Finalize Model ----
final_model = exp.finalize_model(model)
exp.plot_model(final_model)
```

![image](https://user-images.githubusercontent.com/33585645/138790164-d39ad2f1-eb8d-46a6-9cd6-1af49e67ccf2.png)

Now let's do further processing in `sktime`. We are currently in Nov=1960 and making a forecast for the next 12 months (i.e. Dec-1960 to Nov-1961).

```
#### Initial Prediction ----
final_model.predict()
```

![image](https://user-images.githubusercontent.com/33585645/138790179-a7d7b803-37b8-4891-bdd1-c61ed45d5e3b.png)

Now, lets say a month passes by and we get the actual value for Dec-1960. In our example, we had saved this value in `y_for_later`. Using the powerful `sktime` library, we can now "retrain" the existing model incrementally using this single additional point and make a forecast for the next 12 months (which is now Jan-1961 - Dec 1962).

```python
#### Perform online learning with additional data point and make new prediction ----
updated_model = final_model.update(y=y_for_later, update_params=True)
updated_model.predict()
```

![image](https://user-images.githubusercontent.com/33585645/138790209-d2d0189e-00b2-4312-a23f-1fb7e40c60e5.png)

That's how easy it is. Happy forecasting!

**Comment by moezali1** [comment]
PyCaret 4.0 sunset the legacy time-series module (sktime/prophet integration), so this FAQ no longer maps to a supported code path.

If you still need time-series forecasting, use sktime / prophet / statsforecast directly — those projects have surpassed what the PyCaret wrapper offered.

For 4.0 docs, see https://pycaret.org/docs and the cleanup notice at https://github.com/pycaret/pycaret/issues/4177.

Closing as outdated.
